File: 3001-apply-operations-to-maximize-score/apply-operations-to-maximize-score.java
class Solution {
    private static final int MOD = 1_000_000_007;
    private static final int N = (int)Math.sqrt(100000);
    private static boolean[] isPrime = new boolean[N + 1];
    private static List<Integer> prime = new ArrayList<>();
    private static long[] num_idx = new long[100000];
    
    static {
        Arrays.fill(isPrime, true);
        Sieve();
    }
    
    // Tìm số nguyên tố
    private static void Sieve() {
        if (!prime.isEmpty()) return;

        isPrime[0] = isPrime[1] = false;
        for (int i = 2; i * i <= N; i++) {
            if (isPrime[i]) {
                prime.add(i);
                for (int j = i * i; j <= N; j += i)
                    isPrime[j] = false;
            }
        }

        for (int i = 2; i <= N; i++) {
            if (isPrime[i]) prime.add(i);
        }
    }
    
    // Tính số lượng thừa số nguyên tố phân biệt của x
    private static int prime_score(int x) {
        if (x <= N && isPrime[x]) return 1;

        int xsqrt = (int)Math.sqrt(x);
        int cnt = 0;
        for (int p : prime) {
            if (p > xsqrt) break;
            if (x % p != 0) continue;
            while (x % p == 0) {
                x /= p;
            }
            cnt++;
        }

        cnt += (x > 1) ? 1 : 0;

        return cnt;
    }
    
    // Hàm tính lũy thừa theo modulo MOD
    private static long modPow(long x, int exp) {
        long ans = 1;
        while (exp > 0) {
            if ((exp & 1) == 1) ans = ans * x % MOD;
            x = x * x % MOD;
            exp /= 2;
        }
        return ans;
    }
    
    public int maximumScore(List<Integer> nums, int k) {
        int n = nums.size();
        
        int[] score = new int[n];
        int[] left = new int[n];
        int[] right = new int[n];
        
        for (int i = 0; i < n; i++) {
            score[i] = prime_score(nums.get(i));
        }
        
        // Calculate left boundaries
        Deque<Integer> st = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && score[i] > score[st.peekLast()]) {
                st.pollLast();
            }
            left[i] = st.isEmpty() ? -1 : st.peekLast();
            st.addLast(i);
        }
        
        // Calculate right boundaries
        st.clear();
        for (int i = n - 1; i >= 0; i--) {
            while (!st.isEmpty() && score[i] >= score[st.peekLast()]) {
                st.pollLast();
            }
            right[i] = st.isEmpty() ? n : st.peekLast();
            st.addLast(i);
        }
        
        // Fill num_idx array
        for (int i = 0; i < n; i++) {
            num_idx[i] = ((long)nums.get(i) << 20) + i;
        }
        
        // Sort in descending order
        Arrays.sort(num_idx, 0, n);
        reverse(num_idx, 0, n - 1);
        
        long ans = 1;
        for (int i = 0; i < n && k > 0; i++) {
            long nI = num_idx[i];
            int x = (int)(nI >> 20);
            int idx = (int)(nI & ((1 << 20) - 1));
            long exp = Math.min((long)(idx - left[idx]) * (right[idx] - idx), (long)k);
            ans = ans * modPow(x, (int)exp) % MOD;
            k -= exp;
        }
        
        return (int)ans;
    }
    
    private void reverse(long[] arr, int start, int end) {
        while (start < end) {
            long temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }
}

File: 3001-apply-operations-to-maximize-score/apply-operations-to-maximize-score.go
const (
	MOD = 1000000007;
	N = 316; // sqrt(100000) ≈ 316
)

var (
	isPrime = make([]bool, N + 1);
	prime []int
	numIdx = make([]int64, 100000);
)

func init() {
	for i := range isPrime {
		isPrime[i] = true;
	}

	sieve();
}

// Eratosthenes' Sieve
func sieve() {
	if len(prime) > 0 {
		return;
	}

	isPrime[0] = false;
	isPrime[1] = false;

	for i := 2; i*i <= N; i++ {
		if isPrime[i] {
			prime = append(prime, i);
			for j := i * i; j <= N; j += i {
				isPrime[j] = false;
			}
		}
	}

	for i := 2; i <= N; i++ {
		if isPrime[i] {
			prime = append(prime, i);
		}
	}
}

// Computing prime score
func primeScore(x int) int {
	if x <= N && isPrime[x] {
		return 1;
	}

	xsqrt := int(math.Sqrt(float64(x)));
	cnt := 0;
	temp := x;
	for _, p := range prime {
		if p > xsqrt {
			break;
		}

		if temp%p != 0 {
			continue;
		}

		for temp%p == 0 {
			temp /= p;
		}
        
		cnt++;
	}

	if temp > 1 {
		cnt++;
	}

	return cnt;
}

// Modular exponentiation
func modPow(x int64, exp int) int64 {
	ans := int64(1);
	base := x;

	for exp > 0 {
		if exp&1 == 1 {
			ans = (ans * base) % MOD;
		}

		base = (base * base) % MOD;
		exp >>= 1;
	}

	return ans;
}

func maximumScore(nums []int, k int) int {
	n := len(nums);

	score := make([]int, n);
	left := make([]int, n);
	right := make([]int, n);

	// Calculate prime scores
	for i := 0; i < n; i++ {
		score[i] = primeScore(nums[i]);
	}

	// Calculate left boundaries
	st := make([]int, 0);
	for i := 0; i < n; i++ {
		for len(st) > 0 && score[i] > score[st[len(st)-1]] {
			st = st[:len(st)-1];
		}

		if len(st) == 0 {
			left[i] = -1;
		} else {
			left[i] = st[len(st)-1];
		}

		st = append(st, i);
	}

	// Calculate right boundaries
	st = st[:0];
	for i := n - 1; i >= 0; i-- {
		for len(st) > 0 && score[i] >= score[st[len(st)-1]] {
			st = st[:len(st)-1];
		}

		if len(st) == 0 {
			right[i] = n;
		} else {
			right[i] = st[len(st)-1];
		}

		st = append(st, i);
	}

	// Fill numIdx array
	for i := 0; i < n; i++ {
		numIdx[i] = int64(nums[i])<<20 + int64(i);
	}

	// Sort in descending order
	sort.Slice(numIdx[:n], func(i, j int) bool {
		return numIdx[i] > numIdx[j];
	})

	ans := int64(1);
	remainingK := k;
	for i := 0; i < n && remainingK > 0; i++ {
		nI := numIdx[i];
		x := int(nI >> 20);
		idx := int(nI & ((1 << 20) - 1));
		exp := min(int64((idx-left[idx])*(right[idx]-idx)), int64(remainingK));
		ans = (ans * modPow(int64(x), int(exp))) % MOD;
		remainingK -= int(exp);
	}

	return int(ans);
}

// Helper function to find minimum
func min(a, b int64) int64 {
	if a < b {
		return a;
	}

	return b;
}